pkg/fqdn/service/service.go

```diff
@@ -0,0 +1,202 @@
+func (s *FQDNDataServer) OnIPIdentityCacheChange(modType ipcache.CacheModification, cidr types.PrefixCluster, oldHostIP, newHostIP net.IP, oldID *ipcache.Identity, newID ipcache.Identity, encryptKey uint8, k8sMeta *ipcache.K8sMetadata, endpointFlags uint8) {
+	s.identityToIpMutex.Lock()
+	defer s.identityToIpMutex.Unlock()
+
+	ipNet := cidr.AsIPNet()
+
+	if cidr.ClusterID() == 0 && cidr.IsSingleIP() {
+		switch modType {
+		case ipcache.Upsert:
+			if oldID != nil {
+				// Remove from the old identity
+				s.deleteFromIdentityToIPLocked(oldID, ipNet.IP)
+			}
+			s.currentIdentityToIp[newID.ID] = append(s.currentIdentityToIp[newID.ID], ipNet.IP)
+
+		case ipcache.Delete:
+			if oldID != nil {
+				s.deleteFromIdentityToIPLocked(oldID, ipNet.IP)
+			}
+		}
+	}
+}
+
+// deleteFromIdentityToIPLocked deletes the given IP from the identity to IP mapping
+// It is called when the IP identity cache changes and the IP is deleted from the mapping
+// It is also called when the IP is upserted with a new identity
+// It is called with the identityToIpMutex lock held
+func (s *FQDNDataServer) deleteFromIdentityToIPLocked(identity *ipcache.Identity, ip net.IP) error {
+	if identity == nil {
+		return fmt.Errorf("identity is nil")
+	}
+
+	if ips, ok := s.currentIdentityToIp[identity.ID]; ok {
+		newIps := slices.DeleteFunc(ips, func(existing net.IP) bool {
+			return existing.Equal(ip)
+		})
+		if len(newIps) == 0 {
+			delete(s.currentIdentityToIp, identity.ID)
+		} else {
+			s.currentIdentityToIp[identity.ID] = newIps
+		}
+	}
+	return nil
+}
+
+// UpdatePolicyRulesLocked updates the current state of the DNS rules with the given policies and sends the current state of the DNS rules to the client
+// This method is called:
+// 1. when the DNS rules are updated during the endpoint regeneration, we store the state of the DNS rules with flag rulesUpdate as true
+// 2. when the client subscribes to DNS policies, we send the current state of the DNS rules to the client(flag rulesUpdate as false)
+// 3. when the IP identity cache changes, we update the current state of the identity to IP mapping and send the current state of the DNS rules to
+// the client(flag rulesUpdate as false)
+// The UpdatePolicyRulesLocked method is called with the proxy.mutex lock held
+func (s *FQDNDataServer) UpdatePolicyRulesLocked(policies map[identity.NumericIdentity]policy.SelectorPolicy, rulesUpdate bool) error {
+	s.snapshotMutex.Lock()
+	defer s.snapshotMutex.Unlock()
+
+	return nil
+}
+
+// UpdateMappingRequest updates the FQDN mapping with the given data
+// SDP sends the fqdn mapping to cilium agent
+// Steps to update the mapping:
+// 1. Get the endpoint from the IP
+// 2. If the endpoint is not found, return an error
+// 3. If the IPs are not empty, update the cilium agent with the mapping
+func (s *FQDNDataServer) UpdateMappingRequest(ctx context.Context, mappings *pb.FQDNMapping) (*pb.UpdateMappingResponse, error) {
+	return &pb.UpdateMappingResponse{
```

Comment:
The comment talks about updating mappings but this just returns a protobuf message. Also does this need to be exported?

Reply:
I added the note that this is left like this intentionally and will updated in the future PR.
It is one of the method we have from `pb.UnimplementedFQDNDataServer` so that is why exported.